### ISSUE TITLE
Multiple calls make it hang

### ISSUE THREAD

**Issue by jonyo** [felixfbecker/vscode-php-debug#164]
For a project, we have multiple micro services, running in different containers each set up with it's own xdebug port so I can debug a specific micro service or multiple microservices at once.  Everything is working well, breakpoints break, code is able to step properly etc.  Unless one microservice makes a call to another one that makes a call back to the first micro service, at which point it hangs until the connection times out.

Before it times out though, in the editor it shows 2 requests on the call stack.  If I put a breakpoint in the code in "container A" before the point it makes that call to container B, it breaks just fine.  If I then try to step through it, it hangs once it gets to the line that makes the call to container B until the container B call times out, then it continues on with a timeout error creating a stop point in xdebug.

I was able to simplify the error down to a simple test not involving multiple containers or anything, basically just have a call make a curl connection to make a second call.

PHP version:  5.6.30
XDebug version:  2.5.0
Adapter version:  1.10.0

Your launch.json:  
```
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "Listen for XDebug",
			"type": "php",
			"request": "launch",
			"port": 9002,
			"localSourceRoot": "${workspaceRoot}",
			"serverSourceRoot": "/var/www/public/releases/local_source"
		}
	]
}
```
XDebug php.ini config:  
```
zend_extension=xdebug.so
[xdebug]
xdebug.remote_enable=1
xdebug.remote_autostart=1
xdebug.remote_connect_back=0
xdebug.remote_mode=req
xdebug.remote_port=9002
xdebug.remote_host=10.55.55.1
```

XDebug logfile (from setting `xdebug.remote_log` in php.ini):  
```
Log opened at 2017-05-15 17:02:39
I: Connecting to configured address/port: 10.55.55.1:9002.
I: Connected to client. :-)
-> <init xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" fileuri="file:///var/www/public/releases/local_source/app/webroot/test2.php" language="PHP" xdebug:language_version="5.6.30-1+deb.sury.org~xenial+1" protocol_version="1.0" appid="26056"><engine version="2.5.0"><![CDATA[Xdebug]]></engine><author><![CDATA[Derick Rethans]]></author><url><![CDATA[http://xdebug.org]]></url><copyright><![CDATA[Copyright (c) 2002-2016 by Derick Rethans]]></copyright></init>

<- breakpoint_list -i 1
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="1"></response>

<- breakpoint_set -i 2 -t line -f file:///var/www/public/releases/local_source/app/webroot/test2.php -n 10
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="2" id="260560003"></response>

<- breakpoint_list -i 3
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="3"><breakpoint type="line" filename="file:///var/www/public/releases/local_source/app/webroot/test2.php" lineno="10" state="enabled" hit_count="0" hit_value="0" id="260560003"></breakpoint></response>

<- breakpoint_list -i 4
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="4"><breakpoint type="line" filename="file:///var/www/public/releases/local_source/app/webroot/test2.php" lineno="10" state="enabled" hit_count="0" hit_value="0" id="260560003"></breakpoint></response>

<- breakpoint_set -i 5 -t exception -x *
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="5" id="260560004"></response>

<- run -i 6
Log opened at 2017-05-15 17:02:39
I: Connecting to configured address/port: 10.55.55.1:9002.
I: Connected to client. :-)
-> <init xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" fileuri="file:///var/www/public/releases/local_source/app/webroot/test2.php" language="PHP" xdebug:language_version="5.6.30-1+deb.sury.org~xenial+1" protocol_version="1.0" appid="26052"><engine version="2.5.0"><![CDATA[Xdebug]]></engine><author><![CDATA[Derick Rethans]]></author><url><![CDATA[http://xdebug.org]]></url><copyright><![CDATA[Copyright (c) 2002-2016 by Derick Rethans]]></copyright></init>

<- breakpoint_list -i 1
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="1"></response>

<- breakpoint_set -i 2 -t line -f file:///var/www/public/releases/local_source/app/webroot/test2.php -n 10
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="2" id="260520005"></response>

<- stop -i 3
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="stop" transaction_id="3" status="stopped" reason="ok"></response>

-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="run" transaction_id="6" status="stopping" reason="ok"></response>

<- breakpoint_list -i 7
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="breakpoint_list" transaction_id="7"><breakpoint type="line" filename="file:///var/www/public/releases/local_source/app/webroot/test2.php" lineno="10" state="enabled" hit_count="0" hit_value="0" id="260560003"></breakpoint><breakpoint type="exception" state="enabled" hit_count="0" hit_value="0" id="260560004"></breakpoint></response>

<- stop -i 8
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="stop" transaction_id="8" status="stopped" reason="ok"></response>

Log closed at 2017-05-15 17:02:48
```
Adapter logfile (from setting `"log": true` in launch.json):  
```
<- launchResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 2,
  command: 'launch',
  success: true }
new connection 1
<- threadEvent
ThreadEvent {
  seq: 0,
  type: 'event',
  event: 'thread',
  body: { reason: 'started', threadId: 1 } }
<- initializedEvent
InitializedEvent { seq: 0, type: 'event', event: 'initialized' }
-> threadsRequest
{ command: 'threads', type: 'request', seq: 3 }
<- threadsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 3,
  command: 'threads',
  success: true,
  body: { threads: [ Thread { id: 1, name: 'Request 1 (1:02:39 PM)' } ] } }
-> setBreakpointsRequest
{ command: 'setBreakpoints',
  arguments: 
   { source: 
      { path: '/Users/jonathanfoote/zumba/git/public/app/webroot/test2.php',
        name: 'test2.php' },
     lines: [ 10 ],
     breakpoints: [ { line: 10 } ],
     sourceModified: false },
  type: 'request',
  seq: 4 }
<- setBreakpointsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 4,
  command: 'setBreakpoints',
  success: true,
  body: { breakpoints: [ { verified: true, line: 10 } ] } }
-> setFunctionBreakpointsRequest
{ command: 'setFunctionBreakpoints',
  arguments: { breakpoints: [] },
  type: 'request',
  seq: 5 }
<- setFunctionBreakpointsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 5,
  command: 'setFunctionBreakpoints',
  success: true,
  body: { breakpoints: [] } }
-> setExceptionBreakpointsRequest
{ command: 'setExceptionBreakpoints',
  arguments: { filters: [ '*' ] },
  type: 'request',
  seq: 6 }
<- setExceptionBreakpointsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 6,
  command: 'setExceptionBreakpoints',
  success: true }
-> configurationDoneRequest
{ command: 'configurationDone', type: 'request', seq: 7 }
-> threadsRequest
{ command: 'threads', type: 'request', seq: 8 }

<- threadsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 8,
  command: 'threads',
  success: true,
  body: { threads: [ Thread { id: 1, name: 'Request 1 (1:02:39 PM)' } ] } }
new connection 2
<- threadEvent
ThreadEvent {
  seq: 0,
  type: 'event',
  event: 'thread',
  body: { reason: 'started', threadId: 2 } }
<- initializedEvent
InitializedEvent { seq: 0, type: 'event', event: 'initialized' }
-> threadsRequest
{ command: 'threads', type: 'request', seq: 9 }
<- threadsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 9,
  command: 'threads',
  success: true,
  body: 
   { threads: 
      [ Thread { id: 1, name: 'Request 1 (1:02:39 PM)' },
        Thread { id: 2, name: 'Request 2 (1:02:39 PM)' } ] } }
-> setBreakpointsRequest
{ command: 'setBreakpoints',
  arguments: 
   { source: 
      { path: '/Users/jonathanfoote/zumba/git/public/app/webroot/test2.php',
        name: 'test2.php' },
     lines: [ 10 ],
     breakpoints: [ { line: 10 } ],
     sourceModified: false },
  type: 'request',
  seq: 10 }
-> disconnectRequest
{ command: 'disconnect',
  arguments: { restart: false },
  type: 'request',
  seq: 11 }
<- configurationDoneResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 7,
  command: 'configurationDone',
  success: true }
<- outputEvent
OutputEvent {
  seq: 0,
  type: 'event',
  event: 'output',
  body: { category: 'console', output: 'connection 1 closed\n' } }
connection 1 closed
<- threadEvent
ThreadEvent {
  seq: 0,
  type: 'event',
  event: 'thread',
  body: { reason: 'exited', threadId: 1 } }
-> setFunctionBreakpointsRequest
{ command: 'setFunctionBreakpoints',
  arguments: { breakpoints: [] },
  type: 'request',
  seq: 12 }
```

Code snippet to reproduce:  
```php
<?php

if (empty($_GET['connect'])) {
	$ch = curl_init('http://localhost/' . $_SERVER['REQUEST_URI'] . '?connect=1');
	curl_exec($ch);
	curl_close($ch);
	die ('end of initial call');
}

phpinfo();

```

**Comment by SarnaMC** [comment]
Have the same problem

**Comment by ZurabWeb** [comment]
This is really annoying, the browser tab just hangs and time out if there are parallel requests to the same app. I really hope it's fairly simple to fix...

**Comment by SarnaMC** [comment]
@ZurabWeb yes, me too, but look at issue open time. 15 May. 5 months :(

**Comment by ZurabWeb** [comment]
@SarnaMC I think it didn't get the attention of the author. Maybe with more comments we could draw their attention.

**Comment by felixfbecker** [comment]
I get a notification for every comment and more comments won't fix the issue ;)
I don't have the capacity to look into this and it might not even be a bug in the adapter. If you care about this why not look into it yourself? It's open source ⭐️

**Comment by jonyo** [comment]
@felixfbecker Maybe you already noticed this but to me, it looks like on the second request, it attempts to send the breakpoints, and is hung before it gets a response.  Looking at the code, if I'm reading it correctly it looks like it attempts to send the breakpoints to all connections.  I wonder if it tries to send it to the initial connection which is paused so is not going to respond...

I was going to test this theory but can't seem to get the extension set up for debugging, it keeps telling me `Debug adapter process has terminated unexpectedly` when I try to start the debugger in the VSC extension debug window.

**Comment by felixfbecker** [comment]
That could very well be the problem.

I have not debugged this extension in a long time, the config might have changed. What config did you use exactly? Did you follow the contribution guide?

**Comment by calebporzio** [comment]
@jonyo - I'm working on a project that uses microservices and this is a deal breaker for me too - hope you figure something out.

**Comment by jonyo** [comment]
@felixfbecker 

> What config did you use exactly?

For debugging the extension, used the "launch extension" config:
```
    {
      "name": "Launch Extension",
      "type": "extensionHost",
      "request": "launch",
      "runtimeExecutable": "${execPath}",
      "args": ["--extensionDevelopmentPath=${workspaceRoot}"],
      "sourceMaps": true,
      "outDir": "${workspaceRoot}/out"
    }
```

> Did you follow the contribution guide?

I tried to as close as possible, ran into a few issues I am not sure about:

`code testproject --extensionDevelopmentPath=.`

When I try this it gives some errors:

```
[.fseventsd]: Cannot read file .fseventsd/package.json: EACCES: permission denied, open '.fseventsd/package.json'.
Close
Error[.Spotlight-V100]: Cannot read file .Spotlight-V100/package.json: EACCES: permission denied, open '.Spotlight-V100/package.json'.
Close
Error[.PKInstallSandboxManager-SystemSoftware]: Cannot read file .PKInstallSandboxManager-SystemSoftware/package.json: EACCES: permission denied, open '.PKInstallSandboxManager-SystemSoftware/package.json'.
```

I was able to just open the test project using the menus but not sure if that sets it up correctly.

Here's a quick screencast:

https://screencast.com/t/8pqRSIq9iEBy

**Comment by felixfbecker** [comment]
That sounds very weird to me. What npm/node version are you running? Did you try a git clean fresh npm install?

**Comment by jonyo** [comment]
```
$ node --version
v8.7.0
$ npm --version
5.4.2
```
I tried fresh `npm install` and `typings install` as well.

I wonder if it is something to do with the `outDir`, I get a warning that is deprecated in favor of `outFiles`.

This is on macOS 10.13 High Sierra if that makes a difference.

**Comment by felixfbecker** [comment]
You need to use the Node version that VS Code uses, which is v7.
A good reference is always the CI config:
https://github.com/felixfbecker/vscode-php-debug/blob/ed5a99d0d4a5e6840186c21890b7027901609b2e/.travis.yml#L14

**Comment by jonyo** [comment]
@felixfbecker 
Did that, same result... Then I looked at the rest of the travis config and noticed this line...

https://github.com/felixfbecker/vscode-php-debug/blob/ed5a99d0d4a5e6840186c21890b7027901609b2e/.travis.yml#L22

Ran that, now it's working. 👍  Will report back if I find a solution to the multiple call issue.

**Comment by SarnaMC** [comment]
My configuration (use xdebug)
`{
            "name": "Listen for XDebug",
            "type": "php",
            "log": true,
            "request": "launch",
            "port": 9000
        }`

Last log in Debug Console:
`-> setBreakpointsRequest
{ command: 'setBreakpoints',
  arguments: 
   { source: 
      { path: '/var/www/html/drupal/drupal7/htdocs/sites/all/themes/contrib/my-theme/templates/search/blocks/block--views---exp-search-custom-article-page.tpl.php',
        name: 'block--views---exp-search-custom-article-page.tpl.php' },
     lines: [ 2 ],
     breakpoints: [ { line: 2 } ],
     sourceModified: false },
  type: 'request',
  seq: 14 }`

The path is good, and debugging work on Atom so xdebug works

**Comment by jonyo** [comment]
@SarnaMC I see the same thing.

I started debugging, I didn't get very far as I'm not familiar with debugging VSC extensions, I couldn't figure out where to put breakpoints and have it actually work lol...

What I did figure out, I believe the breakpoint requests are sent ASYNC so I think that just happens to be the "last thing" it sends, it may not necessarily be the thing that hands up.  The last response it gets back before it hands is this one...

```
<- threadsResponse
Response {
  seq: 0,
  type: 'response',
  request_seq: 9,
  command: 'threads',
  success: true,
  body: 
   { threads: 
      [ Thread { id: 1, name: 'Request 1 (1:02:39 PM)' },
        Thread { id: 2, name: 'Request 2 (1:02:39 PM)' } ] } }
```

My current working theory is maybe it just isn't set up to fork the connections, so maybe they aren't truly ASYNC.  It does seem to be keeping track of multiple connections, it just hangs up when trying to communicate with multiple ones at same time.  If I can figure out how / where to add breakpoints (adding them in the main src/* files has no effect, adding them in out/* files has no effect either), I might be able to tell more.

**Comment by SarnaMC** [comment]
@jonyo funny but... after computer restart debug works perfectly. What have J done yesterday is J installed node 7.4.0 instead of 8.* and use command npm run compile. Now it works. Thank you @felixfbecker for your support, you are awesome!
J'm going back to work, J will test it all day

**Comment by RaulGRoque** [comment]
The problem really is the multiple connections hitting it at the same time. Looks like this plugin do not support simultaneous connections.

I couldn't do it with this plugin, but the PHPStorm + Komodo DBGP proxy + some additional configurations in PHPStorm did the job. It works perfectly. If anyone is interested I can explain what I did.

Basically these links:
https://confluence.jetbrains.com/display/PhpStorm/Multi-user+debugging+in+PhpStorm+with+Xdebug+and+DBGp+proxy
https://confluence.jetbrains.com/display/PhpStorm/Simultaneous+debugging+sessions+with+PhpStorm

**Comment by jonyo** [comment]
@RaulGRoque That's the weird thing, looking at the code, it appears to be coded to handle multiple connections.  It keeps track of multiple connections, for whatever reason it just doesn't handle it if multiple connections are happening at once.

@felixfbecker Do you have any tips how to get a breakpoint to work?  Like I mentioned in a previous comment, I've tried adding breakpoints in the main code and in the generated code but neither one seems to make it actually pause at the line.  I'm trying to add a breakpoint to the part where it is initially setting up the connection but it isn't working.  Any help would be appreciated!

**Comment by felixfbecker** [comment]
I would have to debug the code. It should work

**Comment by RaulGRoque** [comment]
@jonyo I only achived debugging multiple connections at once when I increased the number in the debug config of PHPStorm. I also unchecked the External Connections options and the XDebug ones like the image below:

![msc](https://user-images.githubusercontent.com/33990167/33576908-82d44c12-d928-11e7-9d9f-3c8a3f7a56bb.png)

**Comment by felixfbecker** [comment]
@RaulGRoque not to be that guy, but please keep the discussions in this repo around this debugger, not PHPStorm. Anyone can use whatever editor they want, but there are folks here (including me) that get notified on every comment and if it's about configuring PHPStorm then there are forums that are a better place than this ❤️

**Comment by RaulGRoque** [comment]
As a side note, I accomplished debugging the backend application in my scenario putting it on an external XAMPP with a particular xDebug port. With the frontend and backend in separated XAMPPs and xDebug ports, the conflicts didn't happen using this plugin one application at a time. But it became messy and the only peaceful solution was this one that I posted above.

**Comment by felixfbecker** [comment]
One big limitation of XDebug is that the whole protocol is synchronous. So while the program (on one Xdebug TCP connection) is running (I sent a `run` command) I cannot send any other request because I need to wait for the response first, which only comes after the program stopped again.

**Comment by carlosesilva** [comment]
I am also having this same issue when trying to debug phpunit tests
+1 for getting this fixed, pretty please =]

**Comment by gaainf** [comment]
For debugging any php code, which was started with exec/system/passthru functions It will be good to add "multiple connections" option to "Listen for XDebug" configuration in launch.json file.
+1

**Comment by jonagoldman** [comment]
Sometimes I forget that the debug session is already running so I click the green play button again (Start Debugging) and now the only solution is to restart the computer.
This bug is so simple to reproduce that it happens to me every single time I debug.

**Comment by elifiner** [comment]
Looks like a simple configuration change is all that is needed to allow Xdebug to work with multiple sessions. All you need is to set the `max_children` setting to a high enough value. This is now my `launch.json`:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Listen for XDebug",
            "type": "php",
            "request": "launch",
            "port": 9000,
            "pathMappings": {
                "/var/www": "${workspaceRoot}",
            },
            "xdebugSettings": {
                "max_children": 20
            },
        },
    ]
}
```

The `max_children` setting seems to mirror a similar setting in PHPStorm mentioned above and it works flawlessly. I can now run Codeception acceptance tests and debug both the test code and the app code at the same time. YAY!

For reference here's my Xdebug configuration in `php.ini`:

```
zend_extension = xdebug.so
xdebug.remote_enable = 1
xdebug.remote_autostart = 1
xdebug.remote_port = 9000
```

**Comment by jonagoldman** [comment]
@elifiner from me it still behaves the same way as mentioned here https://github.com/felixfbecker/vscode-php-debug/issues/164#issuecomment-373959204, even with the `max_children` setting.

**Comment by bdicioccio** [comment]
@elifiner The max_children setting definitely helps with this issue. The site I am dealing with makes a whole lot of ajax requests so it has been very hard to debug. I see the following in the DEBUG console and am lucky if I get to my breakpoint: command is not availableread ECONNRESETwrite ECONNABORTEDwrite ECONNABORTEDread ECONNRESETread ECONNRESETread ECONNRESETread ECONNRESETread ....

**Comment by jonyo** [comment]
@bdicioccio I think there may be a slight difference in your problem, for you it sounds like the multiple requests are being made by the browser.  In my case, the additional request is made server-side.  I don't see any errors like you do, just the normal debug notices, then it just hangs, it acts like it is getting into some kind of deadlock situation.

Too bad I thought maybe it was finally fixed...

Also, I believe what @jonagoldman said is correct, in theory that should just affect how much data it keeps track of by xdebug.  If that fixes it for you, I wonder if somehow the default it is using for you is too high so it is running out of resources or something, so setting it to 20 is limiting it and keeping it from running out of memory.  Just speculating :)

**Comment by jonagoldman** [comment]
I suspect my problem might be related to CORS. An additional OPTIONS request is always sent before the actual request so on paper two requests are always made. The issue now is why this is causing the debugger to pause like there is a hidden break-point somewhere instead of ignoring the request and pausing on the actual break-point in the second request.

**Comment by elifiner** [comment]
@jonagoldman You're right about what the docs say it does seem to have nothing to do with the number of connections. I have no idea why it works, of if that even has something to do with the problem. The only thing I can think of is that when you have enough space in `max_children` to get all the data at once (for some definition of `all`), the connection to Xdebug frees up to be used by the second process. But that's just conjecture, I haven't looked at the code.

Actually, now that I'm thinking about it, I did notice arrays didn't have all the values in watch since the change, they probably have only 20 values max. Could this be what's going on? Could `max_children` limit the size of data retrieved allowing the the Xdebug connection to be released?

@jonyo I definitely have a server side call that results in a request being opened from within another request. I'm running codeception acceptance tests from command line and the tests make curl calls to my site. So there's a command line PHP script (the test code) and the actual HTTP request (the tested code) running at the same time and I have breakpoints working for both at the same time with the `max_children` change.

**Comment by svict4** [comment]
For anyone else coming across this thread: the answer described [in this SO question](https://stackoverflow.com/a/49207236/905571) fixed my error with multiple requests sent simultaneously freezing up VSCode/php debug.

**Comment by jonyo** [comment]
Regarding the original problem, when the server itself is making the 2nd request using curl...

Its been a while since I tried, so thought I would have another go with debugging.

I added breakpoints at the point it is "writing" and point it is "reading" to/from the connection.  I found a few interesting things:

* Due to the order of things, it can sometimes send multiple commands before getting the response for the first command.  I moved one line around and it fixed that but the problem persisted.
* When debugging, for every message sent, it got a response.  There is no "hanging" going on.
* Rather than having the server make the request, I instead had 2 windows open and added a breakpoint in middle of hello world script, and loaded up both.  It "seems" to work in that scenario so the problem seems unique to when it has a breakpoint on a second request.
* I could be wrong but it seems to be sending the breakpoint refresh way too often.  I tried adding a thing so that while it was still making the request to reload breakpoints, it would not process additional requests to reload breakpoints, that seemed to reduce the number of requests significantly.  Also doing this seems to have made some kind of progress, if I hit "continue" a second time it now shows both requests as being paused on a breakpoint but it is still getting locked up at that point.

I'm still troubleshooting but I'll probably need to take a break soon to get some "actual" work done, just wanted to mention what I found so far in case it makes any sense to you @felixfbecker.  If nothing else and I don't make any further progress I can put in a PR for the one bug fix that properly stops it from trying to send a second command until the first one is done.

**Comment by jonyo** [comment]
Got it working!  I'll make a PR to fix.

So, when you send the run command (or step into, etc), xdebug basically doesn't do anything until it hits a breakpoint or the next line is run or whatever.  So if the code it is running hangs up, say you add a `sleep(30)` and it runs that line...  you won't get a response until that code is done.

In this scenario described, a new connection is being made during that time it is hung up running PHP code.  When there is a new connection, vscode makes requests that basically reset all the breakpoints.  But the way it is coded, it attempts to set the breakpoints on ALL connections and it waits for the results of all those calls.  For the first connection, it ends up adding those calls to the queue, but it never gets to run them because it's waiting on that curl request, which is waiting on breakpoint calls...  So it gets deadlocked.

Long story short: I figured out how to fix it, I'm working on a PR that adds some test coverage as well as fixes it.

**Comment by felixfbecker** [comment]
@jonyo Awesome work!

**Comment by sabman3** [comment]
I know you said this was fixed in the latest release, but I'm not sure it is. I have two ajax calls for a page going to server. I see two process running in the debug window, however sometimes it does stop on the breakpoints while at other times it doesn't. Then on top of that even if I tell it to finish and the processes go away from the debug window, the browser freezes until I stop the debug.

{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Listen for XDebug",
            "type": "php",
            "request": "launch",
            "port": 9000,
            "stopOnEntry": true,
            "xdebugSettings": {
                "max_children": 20
            }
        }        
    ]
}

**Comment by yavgech** [comment]
It now hangs when receiving a second Ajax request. My web page sends Ajax requests continuously.
Downgraded to version 1.12.4 and now it works great again.

**Comment by devnix** [comment]
For me, it hangs too when receiving an AJAX request. Still didn't try to downgrade.

**Comment by Zelfapp** [comment]
Running v 1.3.8.0

Any time there are 2 or more ajax calls php debugging hangs.

![image](https://user-images.githubusercontent.com/11559805/64284291-1fb5a300-cf0e-11e9-9046-8fec4b34ffb3.png)

**Comment by jonyo** [comment]
@Zelfapp By the way that sounds like a different issue, this one was a deadlock that happened when one request made a second request, it was not related to multiple ajax requests.

Also this one was resolved and closed over a year ago, you may get more traction creating a new issue or looking for a more recent issue that involves multiple ajax requests.

I see the confusion though if you only read the issue title it does sound related; since you are not the first to comment on it I'll update the title (if it lets me).

**Comment by edelauna** [comment]
I tried removing the number of breakpoints I was using down to 1 per function and it seemed to help.

**Comment by Fabricjs6** [comment]
@jonyo https://stackdev.space/questions/exploring-ajax-requests-in-codeception-acceptance-testing
